Add new keyboard shortcuts for block settings menu

talldan opened the pull request:
## Description
Related issues - #7433, #71, #2980

Adds some new keyboard shortcuts for the block settings menu:
- <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd> / <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd> - duplicate block
- <kbd>Cmd</kbd>+<kbd>shift</kbd>+<kbd>,</kbd> / <kbd>Ctrl</kbd>+<kbd>shift</kbd>+<kbd>,</kbd> - show/hide sidebar

In order to achieve this a number of changes were made:
- `MenuItem`s are now used to render block settings buttons (since they display shortcuts in the intended way)
- Improved some incorrect aria attributes used on `MenuItem`. MenuItem now defaults to a role of `menuitem` and uses `aria-checked` instead of `aria-pressed` as per example implementation (https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-2/menubar-2.html)
- Move `onDuplicate` and `onRemove` action dispatchers from their components to a HOC (`withBlockSettingsActions`) so that they can be shared by a keyboard shortcut component as well as the buttons themselves.
- Remove `BlockRemoveButton` and `BlockDuplicateButton`, since `MenuItem`s can now be used directly
- Add `BlockSettingsKeyboardShortcuts` component for binding of shortcuts
- Alter the way `BlockSettingsMenu` is rendered so that keyboard shortcuts are available when typing.
- Some changes to `RichText` keyboard event handling. The component was stopping propagation of some keyboard events unnecessarily.

## How has this been tested?
- Tested Chrome, Firefox, Safari, Opera on MacOS. Only known keyboard shortcut conflict is Cmd+Shift+D which is used for 'Bookmark all open tabs' in Chrome and Firefox. This is prevented when a block is focused.
- Tested Chrome, Firefox, Opera, Edge, IE 11 on Windows 10. Ctrl+Shift+D again conflicts with 'bookmark all tabs' and is prevented when a block is focused.

## Screenshots <!-- if applicable -->
![screen shot 2018-07-30 at 3 59 23 pm](https://user-images.githubusercontent.com/677833/43384787-9d8d1a94-9411-11e8-8e40-1e83840eebc3.png)

![keyboard-shortcuts](https://user-images.githubusercontent.com/677833/43385954-bd0053b6-9414-11e8-95c4-298a3f38acb2.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

talldan committed:
Add duplicate shortcut to block settings menu

- Move onDuplicate dispatch from BlockDuplicateButton into a new HOC called
`withBlockSettingsActions`
- Add `shortcuts` as a prop in `withBlockSettingsActions` that is passed to
its wrapped component
- Pass onDuplicate and shortcut through to BlockDuplicateButton
- Add new BlockSettingsKeyboardShortcuts component for handling shortcut
functionality

Make BlockDuplicateButton a dumb component - move HOC logic to withBlockSettingsActions

Add keyboard shortcut for block removal (Cmd/Special/Windows + DEL)

- Move block remove action dispatch from BlockRemoveButton up to withBlockSettingsActions
- Make BlockRemoveButton a dumb component
- Add keyboard shortcut for block removal to BlockSettingsKeyboardShortcuts
- Pass onRemove into KeyboardShortcuts and BlockRemoveButton

Add shortcut for toggling the sidebar

- Add shortcut keys to list of keyboard shortcuts for edit-post
- Add open/close dispatchers to EditorModeKeyboardShortcuts and hook up
- Make BlockInspectorButton display the keyboard shortcut

MenuItem should use aria-checked attribute instead of aria-pressed

See example implementation of a MenuBar here for details:
https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-2/menubar-2.html

Provide a default role for MenuItem component of menuitem

Still allows the role to be overriden for other menuitem types
(i.e. menuitemcheckbox, menuitemradio), but defaults to menuitem.

Also makes the provision of isSelected prop optional since the resulting
aria-checked role should not be used with a menu-item.

Update usages of MenuItem that should not use the default role

Switch menu items in block settings menu that have a shortcut to use MenuItem

Use the MenuItem component which correctly renders the shortcut. The need for
wrapping components BlockDuplicateButton and BlockRemoveButton is minimal, so
remove them.

Remove role props, since MenuItem implements a default

Make all block settings menu items use MenuItem component instead of IconButton

Spread additional props for MenuItem onto rendered component

This allows MenuItem to support additional props like `disabled`

Remove stroke being applied to MenuItem icons

Most icons don't have stroke - this style rule was adding an additional stroke
to SVG icons, making them appear chunky and blurry.

Remove use of !important

Fix lack of left margin on menu items with icons

Fix potential access of undefined property

When removing the block, there's momentarily a null item in the blocks array

Render block settings when typing, but make it hidden

This ensures BlockSettingsKeyboardShortcuts is rendered and shortcuts
work while actively typing

talldan committed:
Rename shortcut keycodes to raw/display to match block settings menu

jasmussen commented:
Very very cool work! Love to see the shortcuts there. A few quick ones:

- Let's do `⌘,` instead of `⌘+,` — it's a pattern to not show a plus after the ⌘ symbol for shortcuts. However we _do_ probably want, on Windows, for it to say `Ctrl+,`.
- Let's do `Del` instead of `DEL`. Should there be a plus? I'm leaning towards no. @karmatosed any thoughts?

Testing the PR the keyboard shortcuts are working well. However ⌘, works anywhere, not just with the block focused — is this desired? @afercia I seem to remember the original purpose of this shortcut was to open the sidebar and set focus there, like a "Skip" link. Do we not need this anymore, and if not, should we remove the block menu item entirely?

If we want ⌘, to work everywhere, we might want to show it as a tooltip on the cog too, like we show ⌘S on Save:

<img width="195" alt="screen shot 2018-07-30 at 10 57 19" src="https://user-images.githubusercontent.com/1204802/43387711-7104a462-93e7-11e8-8638-d16e04db409d.png">

afercia commented:
It's important to consider that keyboard shortcuts shouldn't ever conflict with native shortcuts. For example, `Cmd+d` on mac Chrome is used to bookmark a page. Though it's unlikely that users want to bookmark a page while a block is selected, conflicts should be avoided. I'm pretty sure the equivalent shortcuts for Windows conflict with other native functionalities. Have the new shortcuts been tested also on Windows? With a few exceptions, I guess we should always use two modifiers for the shortcuts. 

Ideally, `Cmd+,` should work only when there's a block selected otherwise the sidebar will show "No block selected" when it appears.

> I seem to remember the original purpose of this shortcut was to open the sidebar and set focus there, like a "Skip" link. Do we not need this anymore, and if not, should we remove the block menu item entirely?

I'd prefer to not assume users want to move focus on the sidebar. I think having the ability to toggle the sidebar from the block is a good thing, but the ability to "jump" to the sidebar should be offered by other means.

abrightclearweb commented:
Just to agree with @afercia about `Cmd+d;` that would be pretty annoying for Chrome users to see bookmarks created unnecessarily.

`Cmd+` triggers the font size increase so I'm not sure that will work as a shortcut.

The only one that works on Windows and Mac as expected is` Cmd+del `/ `CTRL+Del`.

talldan commented:
Overriding browser behaviour is not desirable and unfortunately we also can't use key combinations that result in a printed character. This rules out a lot of simpler options.

**Delete Block** - I'm fairly happy with `Cmd+Del` / `Ctrl+Del`. Note that when a user is not typing in a block but has block(s) selected, the user can also use the delete key on its own to delete blocks. I personally feel that it's ok to make both options available, it feels fairly intuitive.

**Duplicate Block** - `Cmd+d` does indeed result in a bookmark when no block is selected, but `preventDefault` is used to prevent that while a block is selected. Avoiding this might be a good idea, since it is used the same way in all browsers. `Shift+Cmd+d` / `Shift+Ctrl+d` could be an alternative (in a few text editors this is the shortcut for duplicating a line), but in Chrome (and Firefox) that's also 'Bookmark all tabs'. Is that obscure enough that we'd consider overriding it?

**Toggle sidebar** - I felt that being able to use the shortcut while no blocks are selected is actually preferable. The sidebar is still present at that point, so why not? One thing I'd change on second glance is that when no blocks are selected, the 'Document' tab should be the selected tab when opening the sidebar.

jasmussen commented:
> Ideally, Cmd+, should work only when there's a block selected otherwise the sidebar will show "No block selected" when it appears.

This is not actually the case in this branch. Since the auto-tab-selection, if you have no blocks selected and invoke the shortcut, you'll simply see the sidebar. 

> Toggle sidebar - I felt that being able to use the shortcut while no blocks are selected is actually preferable. The sidebar is still present at that point, so why not? One thing I'd change on second glance is that when no blocks are selected, the 'Document' tab should be the selected tab when opening the sidebar using the shortcut.

Not completely sure either way. I agree that in practice this seems to work very nice — and the more shortcuts, the better. Additionally, I agree it would feel weird that a shortcut for a global interface is available only when a block is selected. Though if we keep this shortcut global, I would suggest we add a tooltip also to the button in the editor bar, as noted in https://github.com/WordPress/gutenberg/pull/8279#issuecomment-408795124.

The alternative viewpoint is that _because_ the block more menu has an item called "Show Block Settings", the shortcut key for `meta+,` is shown in context of that specific menu item, and suggests that it is a shortcut to showing the particular _block settings tab_, not just the sidebar overall as is the case right now. If we were to change the behavior to show that tab, not just toggle the sidebar, we'd probably want an additional shortcut to toggle the sidebar overall, right?

I think that so long as we have the interface where the block settings tab is automatically chosen when you select a block (and the sidebar is open), it makes the most sense to have this shortcut be for toggling the sidebar, not the tab. Essentially, because the tab is not behaving as a tab, but more like a breadcrumb.

talldan committed:
displayShortcut removes `+` symbol between shortcuts when the provided key is any single character

talldan committed:
displayShortcut capitalizes provided key (e.g. Del instead of DEL)

talldan committed:
Avoid placing a + between ⌘ and key entirely

talldan committed:
Change duplicate shortcut to Cmd/Ctrl shift D

talldan committed:
Change toggleSidebar shortcut to `cmd/ctrl + shift + ,` to avoid conflict with Chrome settings shortcut

talldan committed:
When toggling the sidebar using the keyboard, open the correct tab based on whether block(s) are selected

talldan committed:
Add keyboard shortcuts to tooltips for closing and opening the settings sidebar

talldan commented:
@jasmussen I've changed the shortcut for duplicate to Cmd+Shift+D.

For the sidebar toggle, I've made the change I've suggested, so that the appropriate tab is opened depending on whether a block is selected. 

I also realised Cmd+, clashes with Chrome's open settings shortcut, so I've changed that to Cmd+Shift+,

I'll try to have a thorough test across platforms of the shortcuts tomorrow.

jasmussen commented:
Love it. This is working great for me! I have nothing more to add here, unless the others on the thread have objections I think this is good for final code review. Nice work!

noisysocks commented on edit-post/components/keyboard-shortcuts/index.js:
```diff
@@ -16,36 +16,52 @@ class EditorModeKeyboardShortcuts extends Component {
 	}
 
+	toggleSidebar( event ) {
+		event.preventDefault();
```
What does this do?

talldan replied:
Added a comment to explain it. The shortcut triggers 'bookmark all open tabs' in chrome. Also, with the devtools open it repositions the devtools pane. So we want to prevent those. Not ideal to override a browser shortcut, but the options for a shortcut here are very limited.

Cmd+D is add bookmark
Cmd+Opt+D is show/hide dock

noisysocks commented on packages/components/src/menu-item/style.scss:
```diff
@@ -33,14 +33,12 @@
 	&:hover svg * {
-		stroke: $dark-gray-900 !important;
-		fill: $dark-gray-900 !important;
+		fill: $dark-gray-900;
```
❤️

jasmussen replied:
These !importants were actually intentional. 

Try installing https://wordpress.org/plugins/dropit/, you'll note that it adds plugin icons up there that are colored. The only way we can reliably recolorize those on mouse hover and in particular on active, is by using `!important`. This is unfortunate, but it is also exactly the kind of edgecase that I believe the modifier was designed for.

Perhaps we should add a comment?

jasmussen replied:
This is how it should look:

![how it should work](https://user-images.githubusercontent.com/1204802/43756632-cd47414c-9a15-11e8-8380-823c74a33adf.gif)

This is how it will look if we remove the !important's:

![if importants are removed](https://user-images.githubusercontent.com/1204802/43756636-d2fa9be8-9a15-11e8-9d1d-797b7e8b52db.gif)

Note that the toggled state still works in the Drop It plugin even if the !important's are removed, but this is because Drop it is built by our own Riad so he was careful to make that work. In other words, he added his own important styles in his plugin. But we can't rely on plugin developers to do this, and given SVGs devs will use for plugin icons commonly rely on inline styles, I see no other way.

noisysocks replied:
> Perhaps we should add a comment?
 
👍

talldan replied:
I've added the !important back in for the fill style and added a comment.

I can't add the stroke style, as it makes our dashicons look fuzzy:
![icons-fuzzy](https://user-images.githubusercontent.com/677833/43760855-ea30a6b4-9a55-11e8-840e-4924bfa4cfb7.gif)

jasmussen replied:
That's a great catch. This needs to be scoped better. The only reason those rules exist for the dropdown menu is so as to catch these plugin-added icons:

![screen shot 2018-08-07 at 09 26 56](https://user-images.githubusercontent.com/1204802/43761031-1d7c860a-9a24-11e8-8375-ae015e62e152.png)

There doesn't appear to be any specific rule we can target, alas:

![screen shot 2018-08-07 at 09 27 56](https://user-images.githubusercontent.com/1204802/43761070-36294a76-9a24-11e8-810f-db7a559de7ca.png)

Should we make one?

Alternately, perhaps it's okay to remove the !important rules after all, since it's less important that the hover and active styles in the _more menu_ are colorized than it is in the toolbar.

talldan replied:
Ah - thanks, I didn't spot them there in menu-item form as well.

I can create a separate issue - it does look like a specific way to target the plugins would be the best trade-off.

noisysocks commented on packages/editor/src/components/block-list/block.js:
```diff
@@ -499,7 +499,7 @@ export class BlockListBlock extends Component {
 						clientIds={ clientId }
 						rootClientId={ rootClientId }
-						isHidden={ ! ( isHovered || isSelected ) || hoverArea !== 'right' }
+						isHidden={ ! ( isHovered || isSelected ) || hoverArea !== 'right' || isTypingWithinBlock }
```
🤯

noisysocks commented on packages/editor/src/components/block-settings-menu/with-block-settings-actions.js:
```diff
@@ -64,5 +63,10 @@ export default compose(
-)( BlockDuplicateButton );
+] );
+
+export default withBlockSettingsActions;
```
Can you remind me why we're putting this in its own file, again? Was it just to make `block-settings-menu/index.js` less large?

I'm wondering if it's unnecessary indirection. `withBlockSettingsActions` and `BlockSettingsMenu` are very tightly coupled so it makes sense in my mind that their code remains very near each other.

talldan replied:
Yeah, totally agree. Initially, I tried out using `withBlockSettingsActions` as a HOC for both the `BlockSettingsMenu` and `BlockSettingsKeyboardShortcuts` (to solve the issue where shortcuts weren't being bound until the menu was rendered). 

I agree now that it's only used to wrap one component they can be combined.

talldan replied:
Done in 8db7543.

Have also gone one step further and refactored `BlockSettingsKeyboardShortcuts` into `BlockSettingsMenu` in f847517, as it wasn't doing anything.

noisysocks commented on packages/editor/src/components/block-settings-menu/with-block-settings-actions.js:
```diff
@@ -1,56 +1,55 @@
 			isLocked: !! getTemplateLock( rootClientId ),
+			blocks,
+			canDuplicate,
+			shortcuts,
```
Why have `withSelect` pass `shortcuts` through as a prop here? Would it make sense for `shortcuts` to just be a regular top-level variable, since it never changes when the store updates?

talldan replied:
Now used as just a regular variable since 8db7543

noisysocks reviewed (commented):
Love this—great job! 😍

- Not so sure about <kbd>Cmd</kbd>+<kbd>Del</kbd> as the shortcut for deleting a block. It doesn't work for paragraph blocks because <kbd>Cmd</kbd>+<kbd>Del</kbd> deletes the current line of text in macOS. In most of our non-paragraph blocks, <kbd>Del</kbd> already deletes the block when it is selected.
- Selecting multiple blocks and then pressing <kbd>Cmd</kbd>+<kbd>Del</kbd> doesn't seem to work.
- Unrelated, but should we just kill <kbd>Shift</kbd>+<kbd>Opt</kbd>+<kbd>Cmd</kbd>+<kbd>M</kbd>? It's such an awkward shortcut and I'm not sure that switching to the code editor is a very common or useful function.

talldan commented:
Thanks for the review @noisysocks :)

>  It doesn't work for paragraph blocks because Cmd+Del deletes the current line of text in macOS.

Hmm, I don't seem to have that behaviour on my Mac OS - `Cmd+Del` in a paragraph doesn't do anything on master.

> Selecting multiple blocks and then pressing Cmd+Del doesn't seem to work.

This also works fine for me. I wonder what's different about your particular setup.

abrightclearweb commented:
@noisysocks 
> Unrelated, but should we just kill `Shift+Opt+Cmd+M`? It's such an awkward shortcut and I'm not sure that switching to the code editor is a very common or useful function.

I'd be reluctant to lose a shortcut for switching between visual and code. Not saying it has to be **that** key combo but I understand that some users with a visual impairment write posts in HTML. It would be bad if they were denied that choice because the shortcut didn't exist.

afercia commented:
Maybe `Shift+Opt+Cmd+M` could be simplified but it's an essential shortcut for pro keyboard users and assistive technologies users.

noisysocks commented:
@abrightclearweb @afercia: Makes sense, thanks for explaining the motivation behind that shortcut! 🙂

> I wonder what's different about your particular setup.

We might be getting confused about what the "Delete" key is. It's my fault. I'm referring to this key, which normal people call "Backspace":

![](http://content.provideocoalition.com/uploads/DeleteKeyMacBookPro_smaller.jpg)

This explains the behaviour I saw, though. I was pressing <kbd>Cmd</kbd>+"Backspace" to delete the block because that's normally what the shortcut for deletion is on Mac. Most Mac users don't have a dedicated "Delete" key unless they're using an external keyboard.

I think our best bet is to just cover our bases and allow both `primary( 'del' )` and `primary( 'backspace' )` to delete the block.

talldan commented:
Oh yes, that does make sense. I should've thought of that. I'll bind backspace as well. I supposed we're not too concerned about losing the ability to delete an entire line? It doesn't seem particularly useful outside of coding.

noisysocks commented:
> I supposed we're not too concerned about losing the ability to delete an entire line? It doesn't seem particularly useful outside of coding.

Ah. I use <kbd>Cmd</kbd>+<kbd>Backspace</kbd> all the time when writing 😛

OK, how about we have:

- When editing text: <kbd>Cmd</kbd>+<kbd>Opt</kbd>+<kbd>Backspace</kbd> & <kbd>Cmd</kbd>+<kbd>Opt</kbd>+<kbd>Del</kbd> deletes the selected block.
- When not editing text: <kbd>Backspace</kbd> & <kbd>Del</kbd> deletes the selected block. (This shortcut already exists in `master`.)

talldan commented:
I've just spotted a new issue. The `cmd+del` or `cmd+shift+del` binding isn't triggered in a paragraph block when at the end of a section of text (ie. when there's nothing to delete). The same happens with backspace at the start of a section of text.

Took some time to figure out what it could be. Seems to be this line that's the cause:
https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/rich-text/index.js#L462

Possibly a bit overzealously, the code there is preventing any other keyboard bindings. I'm fairly certain this was working before, but that code has been there for a while, so maybe not.

¯\_(ツ)_/¯

talldan committed:
Make use of stopImmediatePropagation less zealous

Code is supposed to prevent further handling of keyboard events when a block is
deleted, but doesn't consider cases where a block is not deleted:

- If the user hits delete and the block is kept, but content from another block
merged in
- If the user hits delete, but the block has content and is thus not deleted

talldan committed:
Add modifier key exceptions to prevent deletion or merge of blocks

talldan committed:
Make block deletion shortcut Cmd+Opt+Backspace or Cmd+Opt+Delete

- For Windows that's Ctrl+Alt+Backspace or Ctrl+Alt+Delete

talldan commented:
@noisysocks @jasmussen Did some testing. I haven't found any shortcuts we can use comfortably. The best option seems like Cmd+Option+Backspace / Ctrl+Alt+Backspace. I don't think we'll be able to use the Delete key at all.

I'd be interested to see what Ctrl+Alt+Backspace does on a Windows laptop keyboard though, whether the user has to do Fn+Ctrl+Alt+Backspace to trigger the lock screen, or if they provide a workaround. Answers on a postcard.

Some options we can't use:
**Shift + Alt/Option + (Delete | Backspace)**
Shift+Option+Backspace on MacOs in tinyMCE deletes preceding text
Shift+Option+Delete on MacOs in tinyMCE deletes successive text

**Meta+(Delete | Backspace)**
Cmd+Backspace on MacOs deletes preceding text
Ctrl+Backspace on Windows deletes preceding text
Ctrl+Delete on Windows deletes successive text

**Alt/Option+(Delete | Backspace)**
Option+Backspace on MacOs deletes preceding text
Option+Delete on MacOs deletes successive text

**Meta+Option/Alt+(Delete | Backspace)**
Ctrl+Alt+Delete on Windows is the famous shortcut for rebooting/locking/terminating a program

**Meta+Shift+(Delete | Backspace)**
Ctrl+Shift+Delete on Chrome Windows is Clear Browsing Data
Cmd+Shift+Backspace on Chrome Mac is Clear Browsing Data

jasmussen commented:
Meta + Alt + Backspace seems good to me.

talldan committed:
Add further clauses where event should return early

Iron out inconsistent behaviour with use of backspace/delete key when it can
result in deletion or merging of blocks. When specific modifiers are pressed
the event should not have any effect. Here, the use of preventDefault and
stopImmediatePropagation were blocking further binding of keyboard shortcuts
and preventing browser/system level keyboard shortcuts.

talldan committed:
Make keyboard shortcut for delete Cmd+Alt+Delete

talldan committed:
Abbreviate display shortcut to prevent line wrapping in block settings menu

talldan commented on packages/editor/src/components/rich-text/index.js:
```diff
@@ -440,6 +440,12 @@ export class RichText extends Component {
 		) {
+			// These modifiers would not result in modification of text content,
+			// if they're pressed they should not cause merge or deletion of blocks.
+			if ( ( event.metaKey || event.ctrlKey || event.altKey ) && event.shiftKey ) {
```
I feel like these very localised rules about what are actually native keyboard behaviours are a bad idea, and lead to bugs like the one I found yesterday (#8639). The alternative is probably a pretty dramatic refactor though, so shouldn't be carried out on this PR.

If there is a better way, do let me know :)

noisysocks replied:
Not if I'm misunderstanding this, but when I have my cursor at the start of a paragraph block and press <kbd>Shift</kbd>+<kbd>Option</kbd>+<kbd>Backspace</kbd>, it still merges the blocks.

talldan replied:
@noisysocks Ugh, it does look like I made a few false assumptions. I've addressed them here:
e3f5a41e1c39e8c7045c03d8ee50d83216c4f937

The general logic is that the event shouldn't be handled if a modifier + delete/backspace wouldn't result in any text being deleted. Without the early return, some browser/native shortcuts are being swallowed by the RichText block when at the start or end of text content. The event for our delete block shortcut also wasn't being propagated in those cases.

talldan committed:
Merge `withBlockSettingsActions` into `BlockSettingsMenu`

- BlockSettingsMenu was the only consumer of withBlockSettings
- shortcuts object is now not injected into props

talldan committed:
Remove BlockSettingsKeyboardShortcuts component, which had no logic

talldan committed:
Use preventDefault to prevent any additional browser/os shortcuts and add comments

noisysocks commented on edit-post/components/keyboard-shortcuts/index.js:
```diff
@@ -16,36 +16,52 @@ class EditorModeKeyboardShortcuts extends Component {
+		hasBlockSelection: !! select( 'core/editor' ).getBlockSelectionStart(),
+	} ) ),
+	withDispatch( ( dispatch, { hasBlockSelection } ) => ( {
+		switchMode: ( mode ) => {
```
You didn't do this 👀

talldan replied:
Whoops - now done.

noisysocks commented on packages/components/src/menu-item/test/index.js:
```diff
@@ -27,6 +27,22 @@ describe( 'MenuItem', () => {
 				icon="wordpress"
 				isSelected={ true }
+				role="menuitemcheckbox"
+				onClick={ () => {} }
```
Very minor: could use `noop` here

noisysocks commented on packages/editor/src/components/block-settings-menu/index.js:
```diff
@@ -148,6 +184,50 @@ export class BlockSettingsMenu extends Component {
+			}
+		},
+		onRemove() {
+			dispatch( 'core/editor' ).removeBlocks( clientIds );
```
Minor: Using destructuring in our `withSelect` and `withDispatch` callbacks is a nice way to reduce how many times we call `select()` and `dispatch()`. I think it reads a little nicer too—but that's subjective, of course.

```js
withDispatch( ( dispatch, { clientIds, ... } ) => {
	const { insertBlocks, multiSelect, removeBlocks } = dispatch( 'core/editor' );

	return {
		onDuplicate() {
			...
		},
		onRemove() {
			removeBlocks( clientIds );
		},
	};
} )
```

noisysocks reviewed (commented):
Code looks good! This is really close.

~I still can't get <kbd>Cmd</kbd>+<kbd>Opt</kbd>+<kbd>Backspace</kbd> to work most times. It seems to work when I have a paragraph block selected but my cursor not placed within its TinyMCE instance. But when my cursor is placed inside the paragraph block and ready to type, the shortcut doesn't delete the block.~

~I'm also noticing that <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd> doesn't work if we're in "typing mode". To reproduce:~

~1. Insert a paragraph block.~
~2. Start typing some text. The selection outline should go away.~
~3. Press <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd>. Nothing happens.~

Never mind, I wasn't running the latest changes 😅 

There's just a few minor issues. I don't think any of these are blockers for getting this PR merged:

- After you press <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd> on a paragraph block, your cursor is at the start of the block which feels a bit weird to me. I'd expect it to appear in the same position as before or at the end.
- Similarly, after you press <kbd>Cmd</kbd>+<kbd>Option</kbd>+<kbd>Backspace</kbd> on a paragraph block, your cursor moves to the start of the previous block. I'd expect it to appear at the end.
- When I press <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd> on a paragraph block that is selected and has the formatting toolbar shown, I see duplicate controls in the formatting toolbar:  
  <img width="667" alt="lol" src="https://user-images.githubusercontent.com/612155/43757600-9e2efdea-9a5c-11e8-98b7-39547078a39f.png">

noisysocks commented:
Some @iseulde eyes here could be good. I get nervous about changes to the `RichText` event handlers since there's so many edge cases 🙂

talldan committed:
Use object method shorthand

talldan committed:
Add !important back in, with clarifying comment

talldan committed:
Use lodash noop over noop arrow function

talldan committed:
Destructure dispatching functions

talldan commented:
@noisysocks The duplicate toolbar thing is unusual. I couldn't reproduce it - any particular steps that you took?

![no-duplicate-toolbar](https://user-images.githubusercontent.com/677833/43762055-71e54a80-9a59-11e8-9d35-9f36ea1897a0.gif)

Thanks for re-reviewing, I've addressed the latest comments. :)

talldan committed:
Refine exceptions to RichText Backspace/Delete handling

iseulde commented:
I’ll have a look in just a bit!

iseulde commented on packages/editor/src/components/rich-text/index.js:
```diff
@@ -444,22 +444,38 @@ export class RichText extends Component {
+			const isCtrlAltDeletePressed = event.ctrlKey && event.altKey && keyCode === DELETE;
+			const isMetaAltPressed = event.metaKey && event.altKey;
+
+			if ( isMetaShiftDeletePressed || isCtrlShiftDeletePressed || isCtrlAltDeletePressed || isMetaAltPressed ) {
```
Testing on a Mac, `isCtrlShiftDeletePressed` and `isCtrlAltDeletePressed` still delete characters in apps such as Notes. Is this different on Windows? But if we want shortcuts to work for these, we could also just return if any modifier is pressed?

talldan replied:
Hmm, I just tested Notes, TextEdit along with input and textarea elements. Consistently across all of them on the mac Ctrl+Option+Delete and Ctrl+Shift+Delete (just delete and not backspace) doesn't result in any text deletion, and I had the same results on Windows. 

I initially thought about returning early for any modifier, but there are some things that I think we should continue to handle like Cmd+Shift+Backspace (delete preceding text to the start of the line) and Alt+Shift+Backspace (delete preceding word). That these continue to work across multiple paragraph blocks is quite a nice user experience.

iseulde commented:
Why do we need:

> - <kbd>Cmd</kbd>+<kbd>Opt</kbd>+<kbd>Backspace</kbd> / <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>Backspace</kbd> - delete block

If there's just a `RichText` area it's still easy to delete the text or cmd+A and delete the text. If there's no `RichText` area, backspace should work without any modifiers. If there's no merge or replace logic for the block's `RichText` areas, cmd+A should select all text content of that area, and a second cmd+A should select the block (as an intermediate step to select all blocks). Then you can press backspace to delete the block. This last one is not implemented yet, but if so I wonder if this new shortcut  is still useful.

jasmussen commented:
> Why do we need: Cmd+Opt+Backspace / Ctrl+Alt+Backspace - delete block

I _love_ what you mentioned about the consecutive ⌘A selections and allowing to delete. Let's make that happen.

But are we totally sure that we can make that work for blocks provided by plugin authors? I could see someone implementing a full codemirror block, which broke every best practice for our block, was a tab trap and all sorts of other things. Could we enforce ⌘A working there?

talldan committed:
Remove the Remove Block shortcut

talldan committed:
Revert changes to RichText back to master version

talldan commented:
@jasmussen @iseulde I've decided to remove the 'Remove Block' shortcut, and will split it out onto a separate PR. Hopefully that means we can get this merged if we're all happy with the changes and then continue the discussion about Remove Block on the other PR..

Also removed are the changes to the RichText keyDown event handler which I'll incorporate into the other PR.

mtias commented:
I think we need a shortcut for deleting a block, but it's totally fine to split into a separate PR. Thanks a lot for all the work here.

youknowriad commented on packages/components/src/menu-item/index.js:
```diff
@@ -21,7 +21,7 @@ import IconButton from '../icon-button';
  * @return {WPElement} More menu item.
  */
-function MenuItem( { children, className, icon, onClick, shortcut, isSelected = false } ) {
+function MenuItem( { children, className, icon, onClick, shortcut, isSelected, role = 'menuitem', ...props } ) {
```
Minor: I guess if we remove the role explicit prop and just use `role="menuitem"` in JSX, it has the exact same effect. Not sure if it's better or if we want to be explicit.

talldan replied:
My thinking was that `menuitem` would usually be the right role, and at least this way there's some out of the box accessibility. Perhaps it's worth adding something to the docs explaining the preferred usage, and also commenting in the code.

youknowriad reviewed (commented):
Awesome PR, there's probably too much in this PR already. We should try to get this in. The only thing that bothers me is the removal of the duplicate/remove separate components.

youknowriad committed:
Fix unit tests (stylelint)

youknowriad commented on edit-post/keyboard-shortcuts.js:
```diff
@@ -5,7 +5,11 @@ import { rawShortcut, displayShortcut } from '@wordpress/keycodes';
+		display: displayShortcut.secondary( 'm' ),
+	},
+	toggleSidebar: {
+		raw: rawShortcut.primaryShift( ',' ),
```
This shortcut is opening a new tab in chrome (Chrome support page)

talldan replied:
I wonder is this is another keyboard layout issue, as this key combination has no effect on my keyboard (cmd+shift+,).

youknowriad replied:
Probably, I have a french Keyboard

talldan replied:
It looks like there's a `,` key and you hold shift to output `?` on a French keyboard. On a British keyboard layout it's `,` and you hold shift to output `<`.

However, on a British keyboard layout, when you press `cmd+shift+,`, the event handler returns that the `,` was pressed. On the French keyboard layout it returns `?`.

That seems completely inconsistent, and I don't know why that's the case.

afercia commented:
While the implementation of the roles `menuitemcheckbox` and `menuitemradio` is technically correct, actual support is very scarce. JAWS, one of the most advanced screen readers, doesn't support them and it doesn't announce the role. Reference: https://freedomscientific.github.io/VFO-standards-support/aria.html (Last updated 05 June 2018)

NVDA, one of the most advanced screen readers, seems to announce `menuitemradio` but it doesn't announce `menuitemcheckbox` at all. It doesn't even say "menu item". Tested with latest Firefox:

![screenshot 53](https://user-images.githubusercontent.com/1682452/43848480-5286cb02-9b33-11e8-8d21-39eba86e7584.png)

With Safari and VoiceOver, I've noticed a degraded experience with focus management when using the standard VoiceOver keystrokes to activate controls: Control-Option-Spacebar.

Also, the various element with a role "menu" have content that is a bit mixed in its nature. In most of the cases they're equivalent to checkboxes or radio buttons. But, if we want to be strict with ARIA, then all the controls within an element with `role="menu"` should be either menuitem, menuitemcheckbox, or menuitemradio. This is not always the case. For example, the "Copy All Content" button (ClipboardButton) misses a role. It's within a menu but it's not equivalent to a checkbox or radio button. Should it be a menuitem? Is it possible to enforce via code that any item in a menu must have a proper role? Not sure. Also, the span that wraps it should have a `role="presentation"` to ensure the button is perceived by assistive technologies as first child of the menu.
```
<div role="menu" aria-orientation="vertical" aria-labelledby="components-menu-group-label-17">
	<span>
		<button type="button" class=" ... ">Copy All Content</button>
	</span>
</div>
```

Having proper markup with proper ARIA is a bit challenging when it must be done programmatically, and it would probably require an increased level of complexity that should be avoided.

Considering `menuitemcheckbox` and `menuitemradio` support is scarce, I wonder why we would want to use them. Also, if we want to be strict with ARIA then we should implement keyboard navigation properly as described in https://www.w3.org/TR/wai-aria-practices/#menu and in the example linked at the beginning of this PR. Navigation should work only with the arrow keys and all the menuitems should have tabindex="-1" to be not tabbable. I'm not sure we really want to do this. 

It all depends on what we want to communicate to assistive technologies users. Is it important they get the subtle difference between a menu item radio and a menu item checkbox (assuming they work)? Or maybe the only important thing is they get there's a "group" of controls and, for each control, they get its state? I'm at a point where I'm reconsidering the usage of `role="menu"` in the first place, but I'd like to ponder a bit on this. Thankfully, this PR standardizes various things and makes future changes easier 👍 

Sorry for the long comment but I didn't get this PR was changing various things besides "keyboard shortcuts".

talldan commented on packages/editor/src/components/block-settings-menu/index.js:
```diff
@@ -98,31 +122,32 @@ export class BlockSettingsMenu extends Component {
-									role="menuitem"
 								/>
 							) }
-							<BlockDuplicateButton
```
One of the challenges was providing the `withDispatch` props to the KeyboardShortcuts component. Another is that KeyboardShortcuts has to render regardless of whether the settings menu is open or closed (so that the shortcut is always bound and available while in a block).  That meant KeyboardShortcuts couldn't be added to the MenuItems (which would have been ideal).

Because of those constraints KeyboardShortcuts exists in the BlockSettingsMenu, and to avoid duplication I moved the `withDispatch` functions there as well. After that there was very little left to the individual components, they were just functions that returned MenuItems (with a little bit of boolean logic).

@noisysocks did suggest the logic in the dispatch function could become an effect. I've since found another complication though, that BlockSettingsMenu is also rendered in BlockMultiControls and receives different props from that parent component, so that would also have to be considered.

I thought this seemed to be the best trade-off.

youknowriad replied:
Fair enough 👍

talldan commented on packages/components/src/menu-item/index.js:
```diff
@@ -40,7 +40,9 @@ function MenuItem( { children, className, icon, onClick, shortcut, isSelected =
 				icon={ icon }
 				onClick={ onClick }
-				aria-pressed={ isSelected }
+				aria-checked={ isSelected }
```
Yep - I'm relying on an undefined value preventing that attribute from being set on the outputted `button` element.

Both `menuitemcheckbox` and `menuitemradiobutton` are valid roles that can be used with `aria-checked`.

afercia replied:
Yes but please see my last comment. `menuitemcheckbox` and `menuitemradio` support is very scarce. It's pointless to use these roles if they're not communicated to users.

talldan replied:
Would you say it's better to just use `menuitem` in that case? How about `aria-checked`?

noisysocks reviewed (approved):
Let's get this in so that it has time to bake for 3.6. We can iterate on the key combinations and correct the ARIA attributes in follow-up PRs.

Great work on this, @talldan! 👏
